// @dolittle/typescript.build/Distribution/Project/Package.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Dolittle. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const is_valid_path_1 = __importDefault(require("is-valid-path"));
const internal_1 = require("../internal");
/**
 * Represents an npm package
 *
 * @export
 * @class Package
 */
class Package {
    /**
     * Instantiates an instance of {Package}.
     * @param {string} _rootFolder Path to the root of the project containing a package.json file
     * @param {Package} [_parentPackage] The parent {Package} if this {Package} is a yarn workspace
     */
    constructor(rootFolder, _parentPackage) {
        this._parentPackage = _parentPackage;
        if (!is_valid_path_1.default(rootFolder) || !fs_1.default.statSync(rootFolder).isDirectory())
            throw new internal_1.PathIsNotDirectory(rootFolder);
        this.rootFolder = path_1.default.resolve(rootFolder);
        this.path = path_1.default.join(rootFolder, Package.PACKAGE_NAME);
        if (!fs_1.default.existsSync(this.path))
            throw new internal_1.NoPackageJson(this.path);
        this.packageObject = JSON.parse(fs_1.default.readFileSync(this.path));
        const webpackConfigPath = path_1.default.join(this.rootFolder, Package.WEBPACK_CONFIG_NAME);
        if (fs_1.default.existsSync(webpackConfigPath)) {
            this.webpackConfigPath = webpackConfigPath;
        }
    }
    static get PACKAGE_NAME() { return 'package.json'; }
    static get WEBPACK_CONFIG_NAME() { return 'webpack.config.js'; }
    /**
     * Gets the parent package for this yarn workspace
     *
     * @readonly
     */
    get parentPackage() {
        return this._parentPackage;
    }
    /**
     * Whether or not this is yarn workspaces root package
     *
     * @returns
     */
    hasWorkspaces() {
        return this.packageObject.workspaces !== undefined;
    }
    /**
     * Whether or not this package uses webpack
     *
     * @returns
     */
    usesWebpack() {
        return this.webpackConfigPath !== undefined;
    }
}
exports.Package = Package;
