// @dolittle/typescript.build/Distribution/Gulp/Tasks/GulpTasks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const gulp_1 = __importDefault(require("gulp"));
const internal_1 = require("../../internal");
class GulpTasks {
    constructor(_context) {
        this._context = _context;
    }
    get cleanTasks() {
        return internal_1.getCleanTasks(this._context);
    }
    get lintTasks() {
        return internal_1.getLintTasks(this._context);
    }
    get buildTasks() {
        return internal_1.getBuildTasks(this._context);
    }
    get testTasks() {
        return internal_1.getTestTasks(this._context);
    }
    get allTasks() {
        return [...this.cleanTasks.allTasks, ...this.lintTasks.allTasks, ...this.buildTasks.allTasks, ...this.testTasks.allTasks];
    }
}
function getGulpTasks(context) {
    if (GulpTasks.gulpTask === undefined)
        GulpTasks.gulpTask = new GulpTasks(context);
    return GulpTasks.gulpTask;
}
exports.getGulpTasks = getGulpTasks;
function createTask(context, taskName, parallel, createTaskCallback) {
    let task;
    if (context.project.workspaces.length > 0) {
        const workspaceTasks = [];
        context.project.workspaces.forEach(_ => {
            const workspaceTask = createTaskCallback(_);
            workspaceTask.displayName = `${taskName}:${_.workspacePackage.packageObject.name}`;
            workspaceTasks.push(workspaceTask);
        });
        task = parallel ? gulp_1.default.parallel(workspaceTasks) : gulp_1.default.series(workspaceTasks);
        task.displayName = taskName;
    }
    else {
        task = createTaskCallback();
        task.displayName = taskName;
    }
    return task;
}
exports.createTask = createTask;
