// @dolittle/typescript.build/Distribution/Gulp/Tasks/LintTasks.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Dolittle. All rights reserved.
 *  Licensed under the MIT License. See LICENSE in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
const gulp_1 = __importDefault(require("gulp"));
const gulp_tslint_1 = __importDefault(require("gulp-tslint"));
const GulpTasks_1 = require("./GulpTasks");
class LintTasks {
    constructor(_context) {
        this._context = _context;
    }
    /**
     * The task for linting
     *
     * @readonly
     */
    get lintTask() {
        if (this._lintTask === undefined) {
            this._lintTask = this.createLintTask(false);
        }
        return this._lintTask;
    }
    /**
     * The task for linting and fixing
     *
     * @readonly
     */
    get lintFixTask() {
        if (this._lintFixTask === undefined) {
            this._lintFixTask = this.createLintTask(true);
        }
        return this._lintFixTask;
    }
    /**
     * Gets all the tasks
     *
     * @readonly
     */
    get allTasks() {
        return [this.lintTask, this.lintFixTask];
    }
    createLintTask(fix) {
        const task = GulpTasks_1.createTask(this._context, fix ? 'lint-fix' : 'lint', true, workspace => {
            const projectSources = workspace !== undefined ? workspace.sources : this._context.project.sources;
            const tsLintConfigPath = workspace ? workspace.tsLint : this._context.project.tsLint;
            const sourceFiles = projectSources.sourceFiles.sourceFileGlobs.includes.map(_ => _.absolute);
            const excludedSourceFiles = projectSources.sourceFiles.sourceFileGlobs.excludes.map(_ => _.absolute);
            return done => gulp_1.default.src(sourceFiles.concat(excludedSourceFiles.map(_ => '!' + _)))
                .pipe(gulp_tslint_1.default({
                formatter: 'verbose',
                fix,
                configuration: tsLintConfigPath
            }))
                .pipe(gulp_tslint_1.default.report({
                summarizeFailureOutput: true
            }))
                .on('end', done);
        });
        return task;
    }
}
exports.LintTasks = LintTasks;
function getLintTasks(context) {
    if (LintTasks.lintTasks === undefined)
        LintTasks.lintTasks = new LintTasks(context);
    return LintTasks.lintTasks;
}
exports.getLintTasks = getLintTasks;
