// @dolittle/typescript.build/Distribution/Project/Sources/OutputFiles.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/*---------------------------------------------------------------------------------------------
*  Copyright (c) Dolittle. All rights reserved.
*  Licensed under the MIT License. See LICENSE in the project root for license information.
*--------------------------------------------------------------------------------------------*/
const path_1 = __importDefault(require("path"));
const slash_1 = __importDefault(require("slash"));
const internal_1 = require("../../internal");
/**
 * Represents the files that are outputted
 *
 * @export
 * @class OutputFiles
 */
class OutputFiles {
    /**
     * Instantiates an instance of {OutputFiles}.
     * @param {string} _projectRootFolder
     * @param {YarnWorkspace[]} [_workspaces=[]]
     */
    constructor(_projectRootFolder, _workspaces = []) {
        this._projectRootFolder = _projectRootFolder;
        this._workspaces = _workspaces;
        this.root = this._workspaces.length > 0 ? undefined : path_1.default.join(this._projectRootFolder, OutputFiles.FOLDER_NAME);
        this.compiledFilesGlobs = this.createCompiledFileGlobs(...OutputFiles.compiledSourceFilesGlobPatterns);
        this.staticOutputFileGlobs = this.createCompiledFileGlobs(...internal_1.StaticFiles.staticOutputFilesGlobPatterns);
        this.declarationFilesGlobs = this.createCompiledFileGlobs(...OutputFiles.declarationFilesGlobPatterns);
        this.compiledTestsGlobs = this.createCompiledFileGlobs(...OutputFiles.compiledTestFilesGlobPatterns);
    }
    static get FOLDER_NAME() { return 'Distribution'; }
    createCompiledFileGlobs(...globPatterns) {
        const globs = {
            includes: [],
            excludes: []
        };
        if (this._workspaces.length > 0) {
            this._workspaces.forEach(_ => {
                globs.includes.push(...internal_1.createGlobPatterns(_.sources.outputFiles.root, globPatterns, _.sources.outputFiles.root === this._projectRootFolder ? '' : slash_1.default(_.sources.outputFiles.root.replace(`${this._projectRootFolder}${path_1.default.sep}`, ''))));
            });
        }
        else {
            globs.includes.push(...internal_1.createGlobPatterns(this.root, globPatterns, OutputFiles.FOLDER_NAME));
        }
        return globs;
    }
}
exports.OutputFiles = OutputFiles;
OutputFiles.declarationFilesGlobPatterns = ['**/*.d.ts'];
OutputFiles.compiledSourceFilesGlobPatterns = ['**/*.js'];
OutputFiles.compiledTestFilesGlobPatterns = ['**/for_*/**/!(given)/*.js', '**/for_*/*.js'];
